src/realm/util/compression.cpp

```diff
@@ -128,13 +130,35 @@ void grow_arena(compression::CompressMemoryArena& compress_memory_arena)
     compress_memory_arena.resize(n); // Throws
 }
 
+uint8_t read_byte(NoCopyInputStream& is, Span<const char>& buf)
+{
+    if (!buf.size())
+        buf = is.next_block();
+    if (buf.size()) {
+        char c = buf.front();
+        buf = buf.sub_span(1);
+        return c;
+    }
+    return 0;
+}
+
+char peek_byte(NoCopyInputStream& is, Span<const char>& buf)
+{
+    if (!buf.size())
+        buf = is.next_block();
+    if (buf.size())
+        return buf.front();
+    return 0;
+}
+
 struct DecompressInputStreamNone final : public NoCopyInputStream {
     DecompressInputStreamNone(NoCopyInputStream& s, Span<const char> b)
         : source(s)
         , current_block(b)
     {
-        if (!current_block.size())
-            current_block = source.next_block();
+        read_byte(s, current_block); // Algorithm
+        read_byte(s, current_block); // Flags
+        peek_byte(s, current_block);
```

Comment:
Why do we have the third peek_byte here? Just to advance the stream?

Reply:
The later code assumes empty current block equals end of stream, and read_byte doesn't advance to the next block when consuming the last byte of the current one. The one remaining use of peak_byte() in the final version has a comment saying this, but it could also be written a lot more clearly.